themes/default/content/blog/converting-full-terraform-programs-to-pulumi/index.md

```diff
@@ -0,0 +1,184 @@
+  </pulumi-chooser> contains the <code>Vault</code> component, converted from the Terraform module.
+
+The latter file contains some TODOs emitted by the converter that must be addressed by the user.
+
+For example, the converter doesn't (yet!) support the `replace` built-in function, so a TODO is left for the user to address:
+
+{{< chooser language "typescript,python,go,csharp" >}}
+{{% choosable language typescript %}}
+
+```typescript
+const alb = new aws.lb.LoadBalancer(`${name}-alb`, {
+    name: notImplemented("replace(var.name_prefix,\"_\",\"-\")"),
+```
+
+{{% /choosable %}}
+{{% choosable language python %}}
+
+```python
+alb = aws.lb.LoadBalancer(f"{name}-alb",
+    name=not_implemented("replace(var.name_prefix,\"_\",\"-\")"),
+```
+
+{{% /choosable %}}
+{{% choosable language go %}}
+
+```go
+alb, err := lb.NewLoadBalancer(ctx, fmt.Sprintf("%s-alb", name), &lb.LoadBalancerArgs{
+    Name: notImplemented("replace(var.name_prefix,\"_\",\"-\")"),
+```
+
+{{% /choosable %}}
+{{% choosable language csharp %}}
+
+```csharp
+var alb = new Aws.Lb.LoadBalancer($"{name}-alb", new()
+{
+    Name = NotImplemented("replace(var.name_prefix,\"_\",\"-\")"),
+```
+
+{{% /choosable %}}
+{{< /chooser >}}
+
+We can fill in an implementation:
+
+{{< chooser language "typescript,python,go,csharp" >}}
+{{% choosable language typescript %}}
+
+```typescript
+const alb = new aws.lb.LoadBalancer(`${name}-alb`, {
+    name: pulumi.output(args.namePrefix).apply(namePrefix => namePrefix.replace("_", "-")),
+```
+
+{{% /choosable %}}
+{{% choosable language python %}}
+
+```python
+alb = aws.lb.LoadBalancer(f"{name}-alb",
+    name=pulumi.Output.from_input(args["namePrefix"]).apply(lambda np: np.replace("_", "-")),
+```
+
+{{% /choosable %}}
+{{% choosable language go %}}
+
+```go
+alb, err := lb.NewLoadBalancer(ctx, fmt.Sprintf("%s-alb", name), &lb.LoadBalancerArgs{
+    Name: args.NamePrefix.ToStringOutput().ApplyT(func(name string) string {
+        return strings.ReplaceAll(name, "_", "-")
+    }).(pulumi.StringOutput),
+```
+
+{{% /choosable %}}
+{{% choosable language csharp %}}
+
+```csharp
+var alb = new Aws.Lb.LoadBalancer($"{name}-alb", new()
+{
+    Name = args.NamePrefix.Apply(name => name.Replace("_", "-"));
+```
+
+{{% /choosable %}}
+{{< /chooser >}}
+
+There are some other TODOs for missing `merge` and `template_file` support, which also need to be filled in. After addressing these and some other tweaks to make the code compile, we can now run the Pulumi program with `pulumi up` to provision the Vault installation with Pulumi.
+
+JVP: TODO: Show how to fix `merge` and `template_file`.
+
+The converter has saved us a ton of time, converting over 1,000 lines of Terraform to a modern Pulumi language, with only a small number of manual fix-ups. From here, we can leverage our IDE and compiler to further refactor and improve the code, one of the many benefits of Pulumi!
```

Comment:
Does it make sense here to talk about what's next in terms of importing the state to Pulumi or do we want a separate blog post on that when it lands in the converter?

Reply:
I'd leave that to when we do the state converter. Maybe a small comment about running import manually?

Reply:
Luke also was wondering about this as it's likely a common question we'll get. I added a section on it at the end on what can be done today and teasing built-in support coming soon.